src/Persistence/Templates/ControlFactory.cs

```diff
@@ -15,21 +15,21 @@ public ControlFactory(IControlTemplateStore controlTemplateStore)
         _controlTemplateStore = controlTemplateStore ?? throw new ArgumentNullException(nameof(controlTemplateStore));
     }
 
-    public Control Create(string name, string template, ControlPropertiesCollection? properties = null, Control[]? controls = null)
+    public Control Create(string name, string template, ControlPropertiesCollection? properties = null, IEnumerable<Control>? controls = null)
     {
         if (_controlTemplateStore.TryGetTemplateByName(template, out var controlTemplate))
         {
             return new BuiltInControl(name, controlTemplate)
             {
                 Properties = properties ?? new(),
-                Controls = controls ?? Array.Empty<Control>()
+                Controls = (Control[])(controls ?? Array.Empty<Control>())
             };
         }
 
         return new CustomControl(name, new ControlTemplate(template))
         {
             Properties = properties ?? new(),
-            Controls = controls ?? Array.Empty<Control>()
+            Controls = (Control[])(controls ?? Array.Empty<Control>())
```

Comment:
I suspect that cast might fail if `controls` is something different than array. This should be safer:
`(controls ?? Array.Empty<Control>()).ToArray()`

Reply:
your suggestion will create a new copy of array

Reply:
if `controls` is a List or any other collection type, the explicit cast to `Control[]` will throw. If we want to reduce allocations then we should check if it's an array before casting. Something like this:
```
Control[] tmpArray = null;
 if (controls is Control[] arr)
     tmpArray = arr;
 else
     tmpArray = (controls ?? Array.Empty<Control>()).ToArray();
 ```